### DISCUSSION TITLE
--[Request/Need help implenting]-- I want to add a virtual keyboard interface access button

### DISCUSSION THREAD

**Discussion by FunnyCorgi** [copy/v86#1164] [Q&A]
I think this would be good for the UI because, while it exists on mobile, not all computers have every key(e.g. pgup, pgdown, home, end, all f-keys). I think this could be a button that adds a separate section(maybe to the right of the running info section).

If you want, I can open a PR for this. I just need to know how to make each button interface with the virtual keyboard. Also, I would like help with making this so it doesn't link into the keyboard inputs unless the user has it toggled on or has pressed the button to turn it on or something.

**Comment by chschnell** [comment]
Injecting synthetic key strokes into V86 is possible by sending scancode bus events of type `keyboard-code` to the keyboard bus `V86.keyboard_adapter.bus`. A scancode is an integer representing the location of a key on a keyboard.

If you only want to send non-alphanumeric keys (Shift, Ctrl, Alt, ESC, F1-F12, etc.) then it's straight forward, because these keys have fixed scancodes (locations) on all keyboards from all countries for decades (unlike the alpanumeric keys like "A", "B", etc.).

To find the scancode of a specific key, first see [here](https://www.w3.org/TR/uievents-code/#key-alphanumeric-functional) and [here](https://www.w3.org/TR/uievents-code/#key-function-section) for the KeyboardEvent.code values of the non-alphanumeric keys.

Then see ~~[here](https://www.rekoba.de/~cs/v86/keyboard-charset/keyboard_event_scancodes.js)~~ [here](https://www.rekoba.de/~cs/v86/demo/include/keyboard_lib.js) for code that maps between KeyboardEvent.code (string) and keyboard-scancode (16-bit integer) values.

Note that in order to inject a full keystroke into V86, you have to send two bus events, one for key-press and another for key-release. The scancode for release is the same as for press but with the most significant bit (0x80) set to 1.

Most scancodes are 8 bit, but if a scancode has 16 bits you first send the most significant 8 bit to the bus, then the least significant 8.

I've put it all together in [Internationalizing Copy and Paste](https://github.com/copy/v86/discussions/1159), you should find all you need in my code (specifically [keyboard_lib.js](https://www.rekoba.de/~cs/v86/demo/include/keyboard_lib.js), at the bottom ~~I've already prototyped the function you're looking for~~ see function `ev_codes_to_scancodes()` for code that does what you want) and comments there.

EDIT: Fixed outdated infos.

**Reply by chschnell** [reply]
I came around and implemented the KeyboardEvent.code => scancode mapping you'd need.

In **[keyboard_lib.js](https://www.rekoba.de/~cs/v86/demo/include/keyboard_lib.js)** you find a single class `VirtualKeyboard`, here is how to use it to inject non-alphanumeric keystrokes into V86.

For example, let's say you have a running V86 instance `emulator` and you want to inject keystrokes `Ctrl+Alt+Del`:

    const scancodes = VirtualKeyboard.ev_codes_to_scancodes(["ControlLeft", "AltLeft", "Delete"]);
    VirtualKeyboard.paste_scancodes(emulator, scancodes);

It will send the following sequence of key events to `emulator`:

    <Ctrl-Down> <Alt-Down> <Del-Down> <Del-Up> <Alt-Up> <Ctrl-Up>

Or the same represented in 16-bit scancodes:

    1D 38 E053 E0D3 B8 9D

It's tested and works. Note that it's not constrained to non-alphanumerics, this would also work:

    const scancodes = VirtualKeyboard.ev_codes_to_scancodes(["ControlLeft", "AltLeft", "KeyA"]);
    VirtualKeyboard.paste_scancodes(emulator, scancodes);

But you cannot inject continuous text like "Hello, world!" this way. Here's how you'd send continuous text using a US-keyboard layout:

    const keyboard = new VirtualKeyboard("kbdus");
    const scancodes = keyboard.text_to_scancodes("Hello, world!");
    VirtualKeyboard.paste_scancodes(emulator, scancodes);

So it's basically a virtual PC keyboard class for V86.

**Comment by FunnyCorgi** [comment]
@chschnell 
> but with the most significant bit (0x80) set to 1.

What do you mean by this?

**Reply by SuperMaxusa** [reply]
In one sentence: the PS/2 controller registers the key hold until you send a scancode(s) to release the key(s).

https://github.com/copy/v86/blob/f92e6b4b55c3f6e5a327b836375df20fedd45fa4/src/browser/main.js#L2239-L2250

https://www.eecg.utoronto.ca/~pc/courses/241/DE1_SoC_cores/ps2/ps2.html#id478856

**Reply by FunnyCorgi** [reply]
OK, so basically I'm just supposed to bitwise OR. Got it.

**Comment by DevScholar** [comment]
Hello everyone, I recently developed a mobile-friendly mouse polyfill and a virtual keyboard. It should be compatible with emulators like v86. However, since I'm not a professional programmer in the field of emulators and this project utilized AI-assisted programming, its quality might not be top-notch. That said, the basic functionality has been implemented, and this virtual keyboard supports complex key combinations like Ctrl+Alt+Delete.
https://github.com/DevScholar/kbdmouse-js
